compiler/options.nim

```diff
@@ -791,41 +791,57 @@ proc findModule*(conf: ConfigRef; modulename, currentModule: string): AbsoluteFi
       result = findFile(conf, m)
   patchModule(conf)
 
+proc skipQuote(input: string; start: int; seps: set[char] = {'"'}): int =
+  result = 0
+  while start+result < input.len and input[start+result] in seps: inc result
+
 proc findProjectNimFile*(conf: ConfigRef; pkg: string): string =
   const extensions = [".nims", ".cfg", ".nimcfg", ".nimble"]
   var
-    candidates: seq[string] = @[]
+    candidates: HashSet[string]
     dir = pkg
     prev = dir
     nimblepkg = ""
+    srcDir = ""
+    finalSrcDir = ""
   let pkgname = pkg.lastPathPart()
   while true:
-    for k, f in os.walkDir(dir, relative = true):
+    for k, f in os.walkDir(dir, relative = false):
       if k == pcFile and f != "config.nims":
-        let (_, name, ext) = splitFile(f)
+        let (currentDir, name, ext) = splitFile(f)
         if ext in extensions:
           let x = changeFileExt(dir / name, ".nim")
           if fileExists(x):
-            candidates.add x
+            candidates.incl x
           if ext == ".nimble":
             if nimblepkg.len == 0:
               nimblepkg = name
-              # Since nimble packages can have their source in a subfolder,
-              # check the last folder we were in for a possible match.
-              if dir != prev:
-                let x = prev / x.extractFilename()
-                if fileExists(x):
-                  candidates.add x
+              const p1 = """$ssrcDir$s=$s"$[skipQuote]$w"$s$[skipQuote]$s"""
+              const p2 = """$ssrcdir$s=$s"$[skipQuote]$w"$s$[skipQuote]$s"""
+              var fs = newFileStream(f, fmRead)
+              var line = ""
+              if not isNil(fs):
+                while fs.readLine(line):
+                  if scanf(line, p1, srcDir) or scanf(line, p2, srcDir):
```

Comment:
* this is subtle logic, should be factored out into its own proc
* this logic seems brittle, it could have false positives and false negatives (eg: what if it's inside a block comment?)

there must be a more robust way to handle this, eg using the `nimble dump --json` I've added for exactly this purpose:

```
nimble dump --json bigints | jq -r .srcDir
src
```
(but you'd use it via `execCmdEx` + parseJson etc)

Reply:
I've seen this suggestion , what i concern is if I can done it own process,I will not call external process todo some task

Reply:
I don't understand your concern. The code you wrote is brittle and complicated, re-doing the job that's supposed to be done by `nimble`. The code I suggest is robust as it simply queries what nimble already knows.
> I will not call external process todo some task

nim itself does this (shelling out to cc). Did you benchmark and saw it made performance significantly worse?

Reply:
I've done a test project, please check https://travis-ci.org/github/bung87/nimble_dump doing 200 time for each kind

Reply:
what commands should we run to run your benchmark? and what performance are you getting with:
* this PR as it is
* this PR with execCmdEx calling `nimble dump --json`

Reply:
`nimble test` or just check the CI job result https://travis-ci.org/github/bung87/nimble_dump/jobs/744025849

Reply:
thanks for writing this benchmark; note that your benchmark is flawed but in a way that, when fixed, further strenghens your argument, you should use epochTime instead of cpuTime since otherwise you'd ignore the time spent in execCmdEx; with execCmdEx I now get:

```
CPU time [s] 1.702131032943726
[OK] execCmdEx nimble dump --json
CPU time [s] 0.009868860244750977
[OK] findProjectNimFile walkFiles
CPU time [s] 0.009277105331420898
[OK] findProjectNimFile walkDir
```

now that we know nimble dump is indeed much slower, the question is whether it matters in practice to justify doing it in this brittle way (your synthetic test was running 200 times). 
I'm open minded on this.

note, this also relates to my question to @dom96 here:
https://github.com/nim-lang/nimble/issues/803#issuecomment-643484339

> the other question I had remains: is nimble dump 100% accurate or can it be wrong if some logic inside foo.nimble overrides some fields (eg src)?

@dom96 is there a lightweight way to get the same as information as `nimble dump --json` via a library call instead of shell call?

Reply:
I'v add test2.nim to that project, you'll see it doesn't match, I'd like nimble has such an api that I can call directly, but am not sure if it's fine compiler thing requires third party package. for current it work and I have no idea what else user will write `srcDir`